lib/src/config_concatenation.cc

```diff
@@ -0,0 +1,248 @@
+        // ConfigConcatenation
+        if (joined.size() > 1 && context.options().get_allow_unresolved()) {
+            return make_resolve_result(move(new_context), make_shared<config_concatenation>(origin(), joined));
+        } else if (joined.empty()) {
+            // we had just a list of optional references using ${?}
+            return make_resolve_result(move(new_context), shared_value {});
+        } else if (joined.size() == 1) {
+            return make_resolve_result(new_context, joined.front());
+        } else {
+            throw config_exception("Bug in the library: resolved list was joined to too many values");
+        }
+    }
+
+    vector<shared_value> config_concatenation::consolidate(vector<shared_value> pieces) {
+        if (pieces.size() < 2) {
+            return pieces;
+        } else {
+            vector<shared_value> flattened;
+            flattened.reserve(pieces.size());
+            for (auto& v : pieces) {
+                if (auto v_concat = dynamic_pointer_cast<const config_concatenation>(v)) {
+                    flattened.insert(flattened.end(), v_concat->pieces().begin(), v_concat->pieces().end());
+                } else {
+                    flattened.push_back(v);
+                }
+            }
+
+            vector<shared_value> consolidated;
+            consolidated.reserve(flattened.size());
+
+            for (auto& v : flattened) {
+                if (consolidated.empty()) {
+                    consolidated.push_back(v);
+                } else {
+                    join(consolidated, v);
+                }
+            }
+
+            return consolidated;
+        }
+    }
+
+    shared_value config_concatenation::concatenate(std::vector<shared_value> pieces) {
+        vector<shared_value> consolidated { consolidate(pieces) };
+        if (consolidated.empty()) {
+            return nullptr;
+        } else if (consolidated.size() == 1) {
+            return consolidated.front();
+        } else {
+            shared_origin merged_origin = simple_config_origin::merge_origins(consolidated);
+            return make_shared<config_concatenation>(merged_origin, consolidated);
+            return {};
+        }
+    }
+
+    // when you graft a substitution into another object,
+    // you have to prefix it with the location in that object
+    // where you grafted it; but save prefixLength so
+    // system property and env variable lookups don't get
+    // broken.
+    shared_value config_concatenation::relativized(std::string prefix) const {
+        vector<shared_value> new_pieces;
+        for (auto& p : _pieces) {
+            new_pieces.push_back(p->relativized(prefix));
+        }
+
+        return make_shared<config_concatenation>(origin(), move(new_pieces));
+    }
+
+    bool config_concatenation::operator==(config_concatenation const& other) const {
+        // note that "origin" is deliberately NOT part of equality
+        return (_pieces == other._pieces);
```

Comment:
This is fine, but not complete. I left out implementing config_value::operator== because it's messy (we probably need to implement config_value::operator== and have it call a virtual helper that children can override).